### PR TITLE
Pd 55 customer service

### PR DESCRIPTION
**Pull request by sbreingan**

### CODE REVIEW

File: src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.Optional;
+
+@RestController
```
Comment (jamesheward) [inline src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java]: not needed, should be explicitly wired in the context

File: src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@Slf4j
+public class CustomerServiceImpl implements CustomerService {
+
+
+    @Autowired
+    @Qualifier("customer-created-events")
+    MessageChannel customerCreatedEventTopic;
```
Comment (jamesheward) [inline src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java]: constructor injection please

File: src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    MessageChannel customerDeletedEventsTopic;
+
+    private final CustomerRepository customerRepository;
+
+    public CustomerServiceImpl(CustomerRepository customerRepository){
+        this.customerRepository = customerRepository;
+    }
+
+    @Override
+    @RequestMapping(value = "/customers/{customerId}", method = RequestMethod.GET)
```
Comment (jamesheward) [inline src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java]: can use GetMapping (and equivalents elsewhere) - its a bit more succinct

File: src/main/java/gov/scot/payments/customer/PaymentValidationRule.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        this.createdBy = createdBy;
+        this.messageTemplate = messageTemplate;
+        this.name = name;
+        this.precedence = precedence;
+        this.rule = rule;
+    }
+
+    @NonNull @Builder.Default private LocalDateTime createdAt = LocalDateTime.now();
+    private String createdBy;
+    private String messageTemplate;
```
Comment (jamesheward) [inline src/main/java/gov/scot/payments/customer/PaymentValidationRule.java]: non null and final for all except createdBy

File: src/main/java/gov/scot/payments/customer/ServicePersistenceProcessor.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    public ServicePersistenceProcessor(ServiceRepository repository){
+        this.repository = repository;
+    }
+
+    //handler for upload queue
+    @StreamListener("persist-service-updated-events")
+    @Transactional
+    public void onServiceUpdates(KStream<String,Service> serviceUpdatedStream){
+        serviceUpdatedStream.foreach( (k,v) -> {
+            System.out.println("Calling on service update");
```
Comment (jamesheward) [inline src/main/java/gov/scot/payments/customer/ServicePersistenceProcessor.java]: remove print line

File: src/main/java/gov/scot/payments/customer/ServiceServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.kafka.support.KafkaNull;
+import org.springframework.messaging.MessageChannel;
+import org.springframework.messaging.support.GenericMessage;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+@RestController
```
Comment (jamesheward) [inline src/main/java/gov/scot/payments/customer/ServiceServiceImpl.java]: same comments as for customer service

File: src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java (hunk 3 above)
Comment (sbreingan) [reply src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java]: I've left this with a TODO - changing to /customers at the top level results in spring complaining about circular mappings - revist this if time to debug.

File: src/main/java/gov/scot/payments/customer/ServiceServiceImpl.java (hunk 6 above)
Comment (sbreingan) [reply src/main/java/gov/scot/payments/customer/ServiceServiceImpl.java]: Similarly as customer service - TODO and will revisit when time to debug the spring debug issues.

File: src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java (hunk 1 above)
Comment (sbreingan) [reply src/main/java/gov/scot/payments/customer/CustomerServiceImpl.java]: Left as TODO for now.

**Pull request merged**